mountpoint-s3/src/data_cache/disk_data_cache.rs

```diff
@@ -0,0 +1,312 @@
+//! Module for the on-disk data cache implementation.
+
+use std::fs;
+use std::io::{ErrorKind, Write};
+use std::ops::RangeBounds;
+use std::path::PathBuf;
+
+use base64ct::{Base64Url, Encoding};
+use bytes::{BufMut, BytesMut};
+use mountpoint_s3_crt::checksums::crc32c;
+use tracing::{error, trace, warn};
+
+use crate::data_cache::DataCacheError;
+
+use super::{BlockIndex, CacheKey, ChecksummedBytes, DataCache, DataCacheResult};
+
+/// On-disk implementation of [DataCache].
+///
+/// TODO: Store checksums on disk, reconstruct as [ChecksummedBytes] using same checksum avoiding recomputation.
+///
+/// TODO: Store additional metadata with each block such as expected S3 key, ETag, etc..
+pub struct DiskDataCache {
+    block_size: u64,
+    cache_directory: PathBuf,
+}
+
+impl DiskDataCache {
+    /// Create a new instance of an [DiskDataCache] with the specified `block_size`.
+    pub fn new(cache_directory: PathBuf, block_size: u64) -> Self {
+        DiskDataCache {
+            block_size,
+            cache_directory,
+        }
+    }
+
+    /// Get the relative path for the given block.
+    fn get_path_for_key(&self, cache_key: &CacheKey) -> PathBuf {
+        let mut path = self.cache_directory.join("v1");
+
+        // An S3 key may be up to 1024 UTF-8 bytes long, which exceeds the maximum UNIX file name length.
+        // Instead, we encode the key and split into 255 character long directory names.
+        let encoded_s3_key = Base64Url::encode_string(cache_key.s3_key.as_bytes());
```

Comment:
Or [`Base64UrlUnpadded`](https://docs.rs/base64ct/latest/base64ct/struct.Base64UrlUnpadded.html)?

Reply:
updated to use unpadded, we don't care about the lost length info since we are not concatenating multiple base64 values together